src/containers/Dashboard.js

```diff
@@ -0,0 +1,163 @@
+            variant='info'
+            size='sm'
+            onClick={this.getRemoteData}
+            style={{ margin: 10 }}
+            active={this.state.display === Display.BACKEND}
+          >
+            Generate data in backend
+          </Button>
+
+          <Button
+            variant='info'
+            size='sm'
+            onClick={this.fetchStoredData}
+            style={{ margin: 10 }}
+            active={this.state.display === Display.DB}
+          >
+            Fetch data from database
+          </Button>
+        </Row>
+        <Container style={{ display: showDashboard }}>
+          <Row style={{ marginBottom: 100 }}>
+            <Col className='col-4'>
+              <Chart type='doughnut' data={this.state.chartData} />
+            </Col>
+            <Col className='col-4'>
+              <Chart type='pie' data={this.state.chartData} />
+            </Col>
+            <Col className='col-4'>
+              <Chart type='bar' data={this.state.chartData} />
+            </Col>
+          </Row>
+          <Row>
+            <Col className='col-4'>
+              <Chart type='line' data={this.state.chartData} />
+            </Col>
+            <Col className='col-4'>
+              <Chart type='polarArea' data={this.state.chartData} />
+            </Col>
+            <Col className='col-4'>
+              <Chart type='horizontalBar' data={this.state.chartData} />
+            </Col>
+          </Row>
+        </Container>
+        <Container className={showError}>
+          <Alert variant='danger'>Oh snap! You got an error!</Alert>
+        </Container>
+        <Container fluid className={showLoading}>
+          <Spinner animation='border' variant='info' className='mr-2' />
+          <span className='pl-1'>Loading...</span>
+        </Container>
+      </Container>
+    );
+  }
+
+  /* Generate random chart data */
+  generateLocalData() {
+    const newdata = Object.assign({}, this.state.chartData);
+    newdata.datasets[0].data = [
+      Math.floor(Math.random() * 100 + 1),
+      Math.floor(Math.random() * 100 + 1),
+      Math.floor(Math.random() * 100 + 1),
+    ];
+    this.setState({ display: Display.FRONTEND });
+    this.setState({ chartData: newdata });
+  }
+
+  /* Generate random chart data from backend*/
+  async getRemoteData() {
+    try {
+      this.setState({ display: Display.LOADING });
+      const newdata = Object.assign({}, this.state.chartData);
+      newdata.datasets[0].data = await generateRemoteData();
+      this.setState({ display: Display.BACKEND });
+      this.setState({ chartData: newdata });
+    } catch (err) {
+      this.setState({ display: Display.ERROR });
+    }
+  }
+
+  /* Fetch random chart data from db*/
+  async fetchStoredData() {
+    try {
+      this.setState({ display: Display.LOADING });
+      const newdata = Object.assign({}, this.state.chartData);
+      newdata.datasets[0].data = await getStoredData();
+      this.setState({ display: Display.DB });
```

Comment:
These could have been unified into a single call to `setState`

Reply:
Agree, Fixed!